### PR TITLE
Use appropriate code fix title for IDE0059 for unnecessary compound a…

### LINKED ISSUE #38507: IDE0059 has misleading auto-fix name / incorrect preview with unnecessary ??= assignment
**Version Used**: 
3.4.0-beta1-19426-03+b18d90790da95b6862def56802fa4249e4868bb4

**Steps to Reproduce**:
```C#
class Program
{
    static void Main() { }

    static object M(object o) => o ??= new object();
}
```

**Expected Behavior**:
IDE0059 triggers about using `??=` instead of `??` here.

**Actual Behavior**:
image

The good news is the auto-fix will correctly substitute the "??" for the "??=".  However, two issues:
1) It names the auto-fix "Use discard _", even though it won't.
2) The preview doesn't show either the old or the new code being transformed.

### LINKED ISSUE #40483: Unnecessary assignment warning which changes the result
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/858406/unnecessary-assignment-warning-which-changes-the-r.html)._

---

```
using System;

class P
{
    public ref bool f(ref bool z, ref bool q)
    {
        z = ref q;
        return ref z;
    }
}

class Q
{
    static void F()
    {
        bool a = true;
        bool b = false;
        ref var r = ref new P().f(ref a, ref b);
        b = true;

Console.WriteLine(r);
    }

static void Main()
    {
        F();
    }
}
```

Message	IDE0059	Unnecessary assignment of a value to 'b'	

If you remove `b = true` then output is False an not True.

---
#### Visual Studio Feedback System on 12/18/2019, 01:48 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Visual Studio Feedback System on 12/18/2019, 03:43 PM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by mavasani**
…ssigment

Fixes #38507
Fixes #40483

### CODE REVIEW

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs [hunk 1]
```diff
@@ -126,18 +126,35 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
...
+                        // for example "return x += M();" OR "=> x ??= new C();"
+                        // If so, we will be replacing this compound assignment with the underlying binary operation.
+                        // For the above examples, it will be "return x + M();" AND "=> x ?? new C();" respectively.
+                        // For these cases, we want to show the title as "Remove redundant assignment" instead of "Use discard _".
+
+                        var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+                        var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+                        var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+                        if (syntaxFacts.IsLeftSideOfAnyAssignment(node) &&
+                            !syntaxFacts.IsLeftSideOfAssignment(node) &&
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: i can't make sense of this check... how can a node both be the left side of any assignment... but not hte left side of an assignment?

Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: IsLeftSideOfAnyAssignment eventually comes down to this check: http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Syntax/SyntaxKindFacts.cs,fed59b31290ca8e2

IsLeftSideOfAssignment checks if it is a simple assignment expression.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: i guess i'm not understanding then.  why does this only aply to compound assignment?  if you had `x = new C()` wouldn't you want to remove this redundant assignment just as muhc as `x ??= new C()`?

Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: For redundant assignment `x = new C()`, code fix changes it to `_ = new C()`, and the current code fix title of `Use discard _` is appropriate.

For redundant compound assignment `x ??= new C()`, code fix changes it to `x ?? new C()`, so the current code fix title of `Use discard _` is misleading. This change adjust this case for a more suitable code fix title.

Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: Note the code fix itself is working fine, as noted in https://github.com/dotnet/roslyn/issues/38507. The title is misleading for compound assignments.

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: i see.  could you instead use `IsCompoundAssignExpression`?  i think that will be much easier to understand :)

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: Note: you might need to update that helper to add CoalesceAssignmentExpression.  Alternatively, you could write it to do node+kind checks to be safe when new assignment exprs are added.

Comment (mavasani) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: > you might need to update that helper to add CoalesceAssignmentExpression

I was worried what might break if I change that helper to add CoalesceAssignmentExpression :-). Let me do so and find out.

File: src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs [hunk 2]
```diff
@@ -7487,5 +7488,58 @@ void M(int i)
...
+    C M2() => new C();
+}
+        </Document>
+    </Project>
+</Workspace>
+";
+
+            using var testWorkspace = TestWorkspace.Create(source);
+            var (_, action) = await GetCodeActionsAsync(testWorkspace, parameters: default);
+            Assert.Equal(FeaturesResources.Remove_redundant_assignment, action.Title);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: i believe there's a helper you can call liek GetExactActionSetOffered...

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/RemoveUnusedParametersAndValues/RemoveUnusedValueAssignmentTests.cs]: Ah, it's: TestExactActionSetOfferedAsync

File: src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedValuesCodeFixProvider.cs]: the answer should ideally be 'nothing'. t he purpose of the helper is for features to say: i want to continue to operate on new compound assignments as they get added :)

File: src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb [hunk 3]
```diff
@@ -1333,6 +1333,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return TryCast(node, ExpressionSyntax).IsLeftSideOfAnyAssignStatement
         End Function
 
+        Public Function IsLeftSideOfCompoundAssignment(node As SyntaxNode) As Boolean Implements ISyntaxFactsService.IsLeftSideOfCompoundAssignment
+            ' VB does not support compound assignments.
+            Return False
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/LanguageServices/VisualBasicSyntaxFactsService.vb]: I'm confused by this.  VB def has compound assignments...

**Pull request merged**